### ISSUE TITLE
cloud.action start raises "got an unexpected keyword argument 'kwargs'"

### ISSUE THREAD

**Issue by multani** [saltstack/salt#29602]
I'm trying to run the following command:

```
$  sudo salt-run cloud.action start instance=myinstance provider=amazon-frankfurt
Exception occurred in runner cloud.action: Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/client/mixins.py", line 337, in low
    data['return'] = self.functions[fun](*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/runners/cloud.py", line 139, in action
    info = client.action(func, cloudmap, instances, provider, instance, kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 466, in action
    return mapper.do_function(provider, fun, kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1511, in do_function
    call='function', kwargs=kwargs
TypeError: start() got an unexpected keyword argument 'kwargs'
```

The `provider` and `instance` are correct relatively to the salt-cloud setup on the machine. The provider is configured as follow (relevant parameters only?)

```
amazon-frankfurt:
  driver: ec2
  location: eu-central-1
  [...]
```

I can provide more details if needed.

**Comment by multani** [comment]
Might also be of some interest:
```
$ salt --versions-report      
Salt Version:
           Salt: 2015.8.3
 
Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: Not Installed
           Mako: 1.0.0
         PyYAML: 3.11
          PyZMQ: 14.4.0
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.2
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed
 
System Versions:
           dist: debian 8.2 
        machine: x86_64
        release: 3.16.0-4-amd64
         system: debian 8.2 
 
$ salt-cloud --versions-report
Salt Version:
            Salt: 2015.8.3
 
Dependency Versions:
 Apache Libcloud: 0.17.0
          Jinja2: 2.7.3
        M2Crypto: Not Installed
            Mako: 1.0.0
          PyYAML: 3.11
           PyZMQ: 14.4.0
          Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
            RAET: Not Installed
         Tornado: 4.2.1
             ZMQ: 4.0.5
            cffi: Not Installed
        cherrypy: Not Installed
        dateutil: 2.2
           gitdb: Not Installed
       gitpython: Not Installed
           ioflo: Not Installed
         libnacl: Not Installed
    msgpack-pure: Not Installed
  msgpack-python: 0.4.2
    mysql-python: 1.2.3
       pycparser: Not Installed
        pycrypto: 2.6.1
          pygit2: Not Installed
    python-gnupg: Not Installed
           smmap: Not Installed
         timelib: Not Installed
 
System Versions:
            dist: debian 8.2 
         machine: x86_64
         release: 3.16.0-4-amd64
          system: debian 8.2
```

**Comment by jfindlay** [comment]
@multani, thanks for the report.

**Comment by pass-by-value** [comment]
Created #30217 to fix this

**Comment by rallytime** [comment]
@multani I was looking into this a little more to see if we could get this closed up, and while @pass-by-value's fix allows passing kwargs to the action funciton (which is awesome), the way you've written your example command did work with the original changes.

I've fixed this in #30691. Basically, `instance` wasn't being used later `action` function, which aslo produced a stacktrace. Also note that you shouldn't use `instance` with `provider`. That will generate an error. My PR cleans up this stacktrace and gives you an error message and returns an empty dict, like so:
```
# salt-run cloud.action stop instance=rally-to-clone provider=linode
[ERROR   ] Either an instance (or list of names) or a provider must be specified, but not both.
{}
```
But, if you remove the `provider` option and specify the `instance` option only, the command is successful:
```
# salt-run cloud.action stop instance=rally-to-clone
linode-config:
    ----------
    linode:
        ----------
        rally-to-clone:
            ----------
            action:
                stop
            state:
                Stopped
            success:
                True
```

**Comment by rallytime** [comment]
Unfortunately, the fix from @pass-by-value for this issue was causing a lot of regressions in typical salt-cloud -a usage, so we've had to revert that change with #31061. 

We'll have to take another crack at fixing this particular issue.

**Comment by viict** [comment]
I think this is affecting me aswell.

As I i'm not using salt-cloud directly I made a small change on "salt/cloud/__init__.py":

The patch, is based on a regressed commit found [here](https://github.com/rallytime/salt/commit/37bfc99ff983b9df64d5069d0df458d765b64559):
```
1520a1521
>                             kws = { 'call': 'action' }
1522c1523
<                                 vm_name, kwargs, call='action'
---
>                                 vm_name, **kws

```

Well, this made things start to work, I know it's not the best fix.

```

salt --versions-report
Salt Version:
           Salt: 2016.3.2
 
Dependency Versions:
           cffi: 1.6.0
       cherrypy: 3.2.2
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: 1.2.3
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.6 (default, Oct 26 2016, 20:30:19)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.4.0
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 4.1.5
 
System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-105-generic
         system: Linux
        version: Ubuntu 14.04 trusty

```

**Comment by cachedout** [comment]
@rallytime Why was this re-opened? Can you chase this down, please?

**Comment by rallytime** [comment]
Yeah, for sure. I think this was re-opened because the original fix for this from @pass-by-value had to be reverted as it was causing problems elsewhere.

**Comment by cachedout** [comment]
K, sounds good. Let us know what you find and thanks. :]

**Comment by rallytime** [comment]
I am unable reproduce this on recent versions of Salt. I tested this at the HEAD of 2016.3 and 2016.11 as well as 2016.3.6 and 2016.11.3.

While I was testing this, I did find a stacktrace that needed to be handled better when you provide both an `instance` and `provider` kwargs, and I have submitted #40735  to fix the stacktraces. The stacktrace looks like this:
```
# salt-run cloud.action start instance=rally-test-1 provider=linode
Exception occurred in runner cloud.action: Traceback (most recent call last):
  File "/root/SaltStack/salt/salt/client/mixins.py", line 357, in low
    data['return'] = self.functions[fun](*args, **kwargs)
  File "/root/SaltStack/salt/salt/runners/cloud.py", line 152, in action
    info = client.action(func, cloudmap, instances, provider, instance, **_filter_kwargs(kwargs))
  File "/root/SaltStack/salt/salt/cloud/__init__.py", line 490, in action
    'Either an instance (or list of names) or a provider must be '
SaltCloudConfigError: Either an instance (or list of names) or a provider must be specified, but not both.
```
And when you remove the `provider` kwarg, the command works as expected:
```
# salt-run cloud.action start instance=rally-test-1
linode-config:
    ----------
    linode:
        ----------
        rally-test-1:
            ----------
            action:
                start
            msg:
                Machine already running
            state:
                Running
            success:
                True
```
Note that if running the `cloud.action` command with salt or salt-call instead of salt-run, #40278 has a similar issue to this one, and will be fixed in the upcoming 2016.11.4 release of Salt.
